src/ray/raylet/node_manager.cc

```diff
@@ -1368,36 +1363,45 @@ void NodeManager::ProcessDisconnectClientMessage(
   // Publish the worker failure.
   auto worker_failure_data_ptr =
       gcs::CreateWorkerFailureData(self_node_id_, worker->WorkerId(), worker->IpAddress(),
-                                   worker->Port(), time(nullptr), intentional_disconnect);
+                                   worker->Port(), time(nullptr), disconnect_type);
   RAY_CHECK_OK(
       gcs_client_->Workers().AsyncReportWorkerFailure(worker_failure_data_ptr, nullptr));
 
   if (is_worker) {
     const ActorID &actor_id = worker->GetActorId();
     const TaskID &task_id = worker->GetAssignedTaskId();
-    // If the worker was running a task or actor, clean up the task and push an
-    // error to the driver, unless the worker is already dead.
+
     if ((!task_id.IsNil() || !actor_id.IsNil()) && !worker->IsDead()) {
+      // If the worker was running a task or actor, clean up the task and push an
+      // error to the driver, unless the worker is already dead.
+      ErrorType error_type = ErrorType::WORKER_DIED;
+      // TODO(rkn): Define this constant somewhere else.
+      std::string type_str;
+      std::ostringstream error_message;
+      if (disconnect_type == rpc::ClientDisconnectType::UNEXPECTED_EXIT) {
+        type_str = "worker_died";
+        error_message << "A worker died or was killed while executing task " << task_id
+                      << ".";
+      }
+      if (disconnect_type == rpc::ClientDisconnectType::PLACEGROUP_REMOVED) {
+        error_type = ErrorType::PLACEMENT_GROUP_ERROR;
+        type_str = "placement_group_error";
+        error_message << "A worker was killed while executing task " << task_id
+                      << " due to placement group removal"
+                      << ".";
+      }
       // If the worker was an actor, it'll be cleaned by GCS.
       if (actor_id.IsNil()) {
         Task task;
         if (local_queues_.RemoveTask(task_id, &task)) {
-          TreatTaskAsFailed(task, ErrorType::WORKER_DIED);
+          TreatTaskAsFailed(task, error_type);
         }
       }
-
-      if (!intentional_disconnect) {
-        // Push the error to driver.
-        const JobID &job_id = worker->GetAssignedJobId();
-        // TODO(rkn): Define this constant somewhere else.
-        std::string type = "worker_died";
-        std::ostringstream error_message;
-        error_message << "A worker died or was killed while executing task " << task_id
-                      << ".";
-        auto error_data_ptr = gcs::CreateErrorTableData(type, error_message.str(),
-                                                        current_time_ms(), job_id);
-        RAY_CHECK_OK(gcs_client_->Errors().AsyncReportJobError(error_data_ptr, nullptr));
-      }
+      // Push the error to driver.
+      const JobID &job_id = worker->GetAssignedJobId();
+      auto error_data_ptr = gcs::CreateErrorTableData(type_str, error_message.str(),
```

Comment:
```suggestion
      const auto &error_data_ptr = gcs::CreateErrorTableData(type_str, error_message.str(),
```

Reply:
Is this the recommended way to declare erro_data_ptr? it is `auto error_data_ptr ` in the code base

Reply:
`const auto &error_data_ptr` can reduce one copy. Actually, I don't quite understand why was `auto error_data_ptr` before.

Reply:
The return value is actually made inside CreateErrorTableData though? I think we should just do const auto error_data_ptr? (also the return value is anyway a pointer)

Reply:
SGTM

Reply:
updated to const auto everywhere!